python/ml4ir/base/model/architectures/transformer.py

```diff
@@ -0,0 +1,135 @@
+import tensorflow as tf
+
+"""
+Implementation of transformer architecture based on tensorflow tutorial
+Reference -> https://www.tensorflow.org/text/tutorials/transformer
+"""
+
+
+def scaled_dot_product_attention(q, k, v, mask=None):
+    """Calculate the attention weights.
+    q, k, v must have matching leading dimensions.
+    k, v must have matching penultimate dimension, i.e.: seq_len_k = seq_len_v.
+    The mask has different shapes depending on its type(padding or look ahead)
+    but it must be broadcastable for addition.
+
+    Parameters
+    -----------
+    q: Tensor
+        query tensor with shape == (..., seq_len_q, depth)
+    k: Tensor
+        key tensor with shape == (..., seq_len_k, depth)
+    v: Tensor
+        value tensor with shape == (..., seq_len_v, depth_v)
+    mask: Tensor
+        Float tensor with shape broadcastable
+        to (..., seq_len_q, seq_len_k). Defaults to None.
+
+    Returns
+    -------
+    output, attention_weights
+    """
+
+    matmul_qk = tf.matmul(q, k, transpose_b=True)  # (..., seq_len_q, seq_len_k)
+
+    # scale matmul_qk
+    dk = tf.cast(tf.shape(k)[-1], tf.float32)
+    scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)
+
+    # add the mask to the scaled tensor.
+    if mask is not None:
+        # NOTE: Mask is used differently in our implementation
+        #       because 1 - actual record and 0 - padded record
+        scaled_attention_logits += ((1. - mask) * -1e9)
+
+    # softmax is normalized on the last axis (seq_len_k) so that the scores
+    # add up to 1.
+    attention_weights = tf.nn.softmax(
+        scaled_attention_logits, axis=-1)  # (..., seq_len_q, seq_len_k)
+
+    output = tf.matmul(attention_weights, v)  # (..., seq_len_q, depth_v)
+
+    return output, attention_weights
+
+
+def point_wise_feed_forward_network(model_dim, feed_forward_dim):
+    return tf.keras.Sequential([
+        tf.keras.layers.Dense(feed_forward_dim, activation='relu'),  # (batch_size, seq_len, feed_forward_dim)
+        tf.keras.layers.Dense(model_dim)  # (batch_size, seq_len, model_dim)
+    ])
+
+
+class MultiHeadAttention(tf.keras.layers.Layer):
```

Comment:
Why we don't use a [standard transformer layer](https://www.tensorflow.org/api_docs/python/tfm/nlp/layers/TransformerEncoderBlock) here?

Reply:
Let me look into this. I believe we didn't have it back when I wrote this.